File: examples/python/trie_examples.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end = True
    
    def search(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_end
    
    def starts_with(self, prefix):
        node = self.root
        for char in prefix:
            if char not in node.children:
                return False
            node = node.children[char]
        return True

# 1. Find Words
def find_words(board, words):
    trie = Trie()
    for word in words:
        trie.insert(word)
    
    result = set()
    def dfs(i, j, node, path):
        if node.is_end:
            result.add(path)
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]):
            return
        char = board[i][j]
        if char not in node.children:
            return
        board[i][j] = '#'
        for di, dj in [(0,1), (1,0), (0,-1), (-1,0)]:
            dfs(i+di, j+dj, node.children[char], path+char)
        board[i][j] = char
    
    for i in range(len(board)):
        for j in range(len(board[0])):
            dfs(i, j, trie.root, "")
    return list(result)

# 2. Longest Word
def longest_word(words):
    trie = Trie()
    for word in words:
        trie.insert(word)
    
    def dfs(node, path):
        result = path
        for char, child in node.children.items():
            if child.is_end:
                candidate = dfs(child, path + char)
                if len(candidate) > len(result) or (len(candidate) == len(result) and candidate < result):
                    result = candidate
        return result
    
    return dfs(trie.root, "")

# 3. Replace Words
def replace_words(dictionary, sentence):
    trie = Trie()
    for root in dictionary:
        trie.insert(root)
    
    def find_root(word):
        node = trie.root
        for i, char in enumerate(word):
            if char not in node.children:
                return word
            node = node.children[char]
            if node.is_end:
                return word[:i+1]
        return word
    
    return " ".join(find_root(word) for word in sentence.split())

# 4. Word Squares
def word_squares(words):
    trie = Trie()
    for word in words:
        trie.insert(word)
    
    def get_words_with_prefix(prefix):
        node = trie.root
        for char in prefix:
            if char not in node.children:
                return []
            node = node.children[char]
        
        result = []
        def dfs(node, path):
            if node.is_end:
                result.append(prefix + path)
            for char, child in node.children.items():
                dfs(child, path + char)
        dfs(node, "")
        return result
    
    def backtrack(square):
        if len(square) == len(words[0]):
            return [square[:]]
        
        prefix = "".join(square[i][len(square)] for i in range(len(square)))
        candidates = get_words_with_prefix(prefix)
        
        result = []
        for word in candidates:
            square.append(word)
            result.extend(backtrack(square))
            square.pop()
        return result
    
    return backtrack([])

# 5. Map Sum
def map_sum(pairs):
    trie = Trie()
    values = {}
    
    def insert(key, val):
        values[key] = val
        node = trie.root
        for char in key:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end = True
    
    def sum_prefix(prefix):
        node = trie.root
        for char in prefix:
            if char not in node.children:
                return 0
            node = node.children[char]
        
        total = 0
        def dfs(node, path):
            nonlocal total
            if node.is_end and prefix + path in values:
                total += values[prefix + path]
            for char, child in node.children.items():
                dfs(child, path + char)
        dfs(node, "")
        return total
    
    for key, val in pairs:
        insert(key, val)
    return sum_prefix

# 6. Palindrome Pairs
def palindrome_pairs(words):
    def is_palindrome(s):
        return s == s[::-1]
    
    result = []
    for i in range(len(words)):
        for j in range(len(words)):
            if i != j:
                combined = words[i] + words[j]
                if is_palindrome(combined):
                    result.append([i, j])
    return result

# 7. Auto Complete System
def auto_complete_system(sentences, times):
    trie = Trie()
    counts = {}
    
    for sentence, time in zip(sentences, times):
        trie.insert(sentence)
        counts[sentence] = time
    
    def get_suggestions(prefix):
        node = trie.root
        for char in prefix:
            if char not in node.children:
                return []
            node = node.children[char]
        
        suggestions = []
        def dfs(node, path):
            if node.is_end:
                suggestions.append((counts.get(prefix + path, 0), prefix + path))
            for char, child in node.children.items():
                dfs(child, path + char)
        dfs(node, "")
        
        suggestions.sort(key=lambda x: (-x[0], x[1]))
        return [s[1] for s in suggestions[:3]]
    
    return get_suggestions

# 8. Stream Checker
def stream_checker(words):
    trie = Trie()
    for word in words:
        trie.insert(word[::-1])
    
    stream = []
    def query(letter):
        stream.append(letter)
        node = trie.root
        for i in range(len(stream) - 1, -1, -1):
            char = stream[i]
            if char not in node.children:
                return False
            node = node.children[char]
            if node.is_end:
                return True
        return False
    
    return query

# 9. Concatenated Words
def concatenated_words(words):
    trie = Trie()
    for word in words:
        trie.insert(word)
    
    def can_form(word, start, count):
        if start == len(word):
            return count > 1
        
        node = trie.root
        for i in range(start, len(word)):
            char = word[i]
            if char not in node.children:
                return False
            node = node.children[char]
            if node.is_end and can_form(word, i + 1, count + 1):
                return True
        return False
    
    return [word for word in words if can_form(word, 0, 0)]

# 10. Word Filter
def word_filter(words):
    """Find words that match both prefix and suffix"""
    trie = Trie()
    
    # Insert all words with special encoding
    for i, word in enumerate(words):
        for j in range(len(word) + 1):
            # Insert suffix + '#' + word
            encoded = word[j:] + '#' + word
            trie.insert(encoded)
    
    def search(prefix, suffix):
        # Search for suffix + '#' + prefix
        query = suffix + '#' + prefix
        node = trie.root
        for char in query:
            if char not in node.children:
                return -1
            node = node.children[char]
        
        # Find the largest index
        max_index = -1
        def dfs(node, path):
            nonlocal max_index
            if node.is_end:
                # Extract original word and find its index
                if '#' in query + path:
                    word = (query + path).split('#')[1]
                    if word in words:
                        max_index = max(max_index, words.index(word))
            for char, child in node.children.items():
                dfs(child, path + char)
        
        dfs(node, "")
        return max_index
    
    return search

# Run Examples
def run_examples():
    print("=== Trie Examples ===")
    
    # Test 1: Basic Trie
    trie = Trie()
    trie.insert("apple")
    print(f"1. Search 'apple': {trie.search('apple')}")
    print(f"   Search 'app': {trie.search('app')}")
    print(f"   Starts with 'app': {trie.starts_with('app')}")
    
    # Test 2: Word Search II
    board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]
    words = ["oath","pea","eat","rain"]
    print(f"2. Find Words: {find_words(board, words)}")
    
    # Test 3: Longest Word
    words = ["w","wo","wor","worl","world"]
    print(f"3. Longest Word: {longest_word(words)}")
    
    # Test 4: Replace Words
    dictionary = ["cat","bat","rat"]
    sentence = "the cattle was rattled by the battery"
    print(f"4. Replace Words: {replace_words(dictionary, sentence)}")
    
    # Test 5: Word Squares
    words = ["area","lead","wall","lady","ball"]
    print(f"5. Word Squares: {word_squares(words)}")
    
    # Test 6: Map Sum
    pairs = [("apple", 3), ("app", 2)]
    sum_func = map_sum(pairs)
    print(f"6. Map Sum 'ap': {sum_func('ap')}")
    
    # Test 7: Palindrome Pairs
    words = ["code","edoc","da","d"]
    print(f"7. Palindrome Pairs: {palindrome_pairs(words)}")
    
    # Test 8: Auto Complete System
    sentences = ["i love you", "island", "iroman", "i love leetcode"]
    times = [5, 3, 2, 2]
    autocomplete = auto_complete_system(sentences, times)
    print(f"8. Auto Complete 'i': {autocomplete('i')}")
    
    # Test 9: Stream Checker
    words = ["cd","f","kl"]
    checker = stream_checker(words)
    print(f"9. Stream Checker 'c': {checker('c')}")
    print(f"   Stream Checker 'd': {checker('d')}")
    
    # Test 10: Concatenated Words
    words = ["cat","cats","catsdogcats","dog","dogcatsdog","hippopotamuses","rat","ratcatdogcat"]
    print(f"10. Concatenated Words: {concatenated_words(words)}")

if __name__ == "__main__":
    run_examples()

File: examples/java/TrieExamples.java
import java.util.*;

public class TrieExamples {
    
    static class TrieNode {
        TrieNode[] children = new TrieNode[26];
        boolean isEnd = false;
    }
    
    // 1. Implement Trie (Prefix Tree)
    static class Trie {
        TrieNode root;
        
        public Trie() {
            root = new TrieNode();
        }
        
        public void insert(String word) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.isEnd = true;
        }
        
        public boolean search(String word) {
            TrieNode node = searchPrefix(word);
            return node != null && node.isEnd;
        }
        
        public boolean startsWith(String prefix) {
            return searchPrefix(prefix) != null;
        }
        
        private TrieNode searchPrefix(String prefix) {
            TrieNode node = root;
            for (char c : prefix.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) return null;
                node = node.children[idx];
            }
            return node;
        }
    }
    
    // 2. Design Add and Search Words Data Structure
    static class WordDictionary {
        TrieNode root;
        
        public WordDictionary() {
            root = new TrieNode();
        }
        
        public void addWord(String word) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.isEnd = true;
        }
        
        public boolean search(String word) {
            return dfs(word, 0, root);
        }
        
        private boolean dfs(String word, int idx, TrieNode node) {
            if (idx == word.length()) return node.isEnd;
            char c = word.charAt(idx);
            if (c == '.') {
                for (TrieNode child : node.children) {
                    if (child != null && dfs(word, idx + 1, child)) return true;
                }
                return false;
            } else {
                int i = c - 'a';
                return node.children[i] != null && dfs(word, idx + 1, node.children[i]);
            }
        }
    }
    
    // 3. Word Search II
    public List<String> findWords(char[][] board, String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.isEnd = true;
        }
        
        Set<String> result = new HashSet<>();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                dfs(board, i, j, root, "", result);
            }
        }
        return new ArrayList<>(result);
    }
    
    private void dfs(char[][] board, int i, int j, TrieNode node, String word, Set<String> result) {
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || board[i][j] == '#') return;
        char c = board[i][j];
        int idx = c - 'a';
        if (node.children[idx] == null) return;
        
        node = node.children[idx];
        word += c;
        if (node.isEnd) result.add(word);
        
        board[i][j] = '#';
        dfs(board, i+1, j, node, word, result);
        dfs(board, i-1, j, node, word, result);
        dfs(board, i, j+1, node, word, result);
        dfs(board, i, j-1, node, word, result);
        board[i][j] = c;
    }
    
    // 4. Replace Words
    public String replaceWords(List<String> dictionary, String sentence) {
        TrieNode root = new TrieNode();
        for (String word : dictionary) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.isEnd = true;
        }
        
        StringBuilder result = new StringBuilder();
        for (String word : sentence.split(" ")) {
            if (result.length() > 0) result.append(" ");
            result.append(findRoot(root, word));
        }
        return result.toString();
    }
    
    private String findRoot(TrieNode root, String word) {
        TrieNode node = root;
        StringBuilder prefix = new StringBuilder();
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (node.children[idx] == null) return word;
            node = node.children[idx];
            prefix.append(c);
            if (node.isEnd) return prefix.toString();
        }
        return word;
    }
    
    // 5. Map Sum Pairs
    static class MapSum {
        TrieNode root;
        
        static class TrieNode {
            TrieNode[] children = new TrieNode[26];
            int val = 0;
        }
        
        public MapSum() {
            root = new TrieNode();
        }
        
        public void insert(String key, int val) {
            TrieNode node = root;
            for (char c : key.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.val = val;
        }
        
        public int sum(String prefix) {
            TrieNode node = root;
            for (char c : prefix.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) return 0;
                node = node.children[idx];
            }
            return dfs(node);
        }
        
        private int dfs(TrieNode node) {
            int sum = node.val;
            for (TrieNode child : node.children) {
                if (child != null) sum += dfs(child);
            }
            return sum;
        }
    }
    
    // 6. Longest Word in Dictionary
    public String longestWord(String[] words) {
        TrieNode root = new TrieNode();
        for (String word : words) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.isEnd = true;
        }
        
        String result = "";
        for (String word : words) {
            if (canBuild(root, word) && (word.length() > result.length() || 
                (word.length() == result.length() && word.compareTo(result) < 0))) {
                result = word;
            }
        }
        return result;
    }
    
    private boolean canBuild(TrieNode root, String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            int idx = c - 'a';
            if (node.children[idx] == null || !node.children[idx].isEnd) return false;
            node = node.children[idx];
        }
        return true;
    }
    
    // 7. Search Suggestions System
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Arrays.sort(products);
        TrieNode root = new TrieNode();
        for (String product : products) {
            TrieNode node = root;
            for (char c : product.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.isEnd = true;
        }
        
        List<List<String>> result = new ArrayList<>();
        TrieNode node = root;
        for (char c : searchWord.toCharArray()) {
            int idx = c - 'a';
            if (node != null && node.children[idx] != null) {
                node = node.children[idx];
                List<String> suggestions = new ArrayList<>();
                dfsProducts(node, searchWord.substring(0, result.size() + 1), suggestions);
                result.add(suggestions.subList(0, Math.min(3, suggestions.size())));
            } else {
                node = null;
                result.add(new ArrayList<>());
            }
        }
        return result;
    }
    
    private void dfsProducts(TrieNode node, String prefix, List<String> suggestions) {
        if (suggestions.size() == 3) return;
        if (node.isEnd) suggestions.add(prefix);
        for (int i = 0; i < 26; i++) {
            if (node.children[i] != null) {
                dfsProducts(node.children[i], prefix + (char)('a' + i), suggestions);
            }
        }
    }
    
    // 8. Palindrome Pairs
    public List<List<Integer>> palindromePairs(String[] words) {
        TrieNode root = new TrieNode();
        for (int i = 0; i < words.length; i++) {
            String word = new StringBuilder(words[i]).reverse().toString();
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.isEnd = true;
        }
        
        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            TrieNode node = root;
            for (int j = 0; j <= word.length(); j++) {
                if (node.isEnd && isPalindrome(word, j, word.length() - 1)) {
                    result.add(Arrays.asList(i, Arrays.asList(words).indexOf(new StringBuilder(word.substring(0, j)).reverse().toString())));
                }
                if (j < word.length()) {
                    int idx = word.charAt(j) - 'a';
                    if (node.children[idx] == null) break;
                    node = node.children[idx];
                }
            }
        }
        return result;
    }
    
    private boolean isPalindrome(String s, int left, int right) {
        while (left < right) {
            if (s.charAt(left++) != s.charAt(right--)) return false;
        }
        return true;
    }
    
    // 9. Word Break II
    public List<String> wordBreak(String s, List<String> wordDict) {
        TrieNode root = new TrieNode();
        for (String word : wordDict) {
            TrieNode node = root;
            for (char c : word.toCharArray()) {
                int idx = c - 'a';
                if (node.children[idx] == null) node.children[idx] = new TrieNode();
                node = node.children[idx];
            }
            node.isEnd = true;
        }
        
        List<String> result = new ArrayList<>();
        dfsWordBreak(s, 0, root, root, "", result);
        return result;
    }
    
    private void dfsWordBreak(String s, int idx, TrieNode root, TrieNode node, String path, List<String> result) {
        if (idx == s.length()) {
            if (node == root) result.add(path.trim());
            return;
        }
        
        int charIdx = s.charAt(idx) - 'a';
        if (node.children[charIdx] != null) {
            TrieNode next = node.children[charIdx];
            if (next.isEnd) {
                dfsWordBreak(s, idx + 1, root, root, path + s.substring(idx - (path.split(" ").length > 0 ? path.split(" ")[path.split(" ").length - 1].length() : 0), idx + 1) + " ", result);
            }
            dfsWordBreak(s, idx + 1, root, next, path, result);
        }
    }
    
    // 10. Stream of Characters
    static class StreamChecker {
        TrieNode root;
        StringBuilder sb;
        
        public StreamChecker(String[] words) {
            root = new TrieNode();
            sb = new StringBuilder();
            for (String word : words) {
                TrieNode node = root;
                for (int i = word.length() - 1; i >= 0; i--) {
                    char c = word.charAt(i);
                    int idx = c - 'a';
                    if (node.children[idx] == null) node.children[idx] = new TrieNode();
                    node = node.children[idx];
                }
                node.isEnd = true;
            }
        }
        
        public boolean query(char letter) {
            sb.append(letter);
            TrieNode node = root;
            for (int i = sb.length() - 1; i >= 0; i--) {
                char c = sb.charAt(i);
                int idx = c - 'a';
                if (node.children[idx] == null) return false;
                node = node.children[idx];
                if (node.isEnd) return true;
            }
            return false;
        }
    }
    
    public static void main(String[] args) {
        TrieExamples examples = new TrieExamples();
        
        System.out.println("=== Trie Examples ===");
        
        // Test 1: Implement Trie
        Trie trie = new Trie();
        trie.insert("apple");
        System.out.println("1. Trie search 'apple': " + trie.search("apple"));
        
        // Test 2: Add and Search Words
        WordDictionary wd = new WordDictionary();
        wd.addWord("bad");
        wd.addWord("dad");
        System.out.println("2. WordDict search 'b.d': " + wd.search("b.d"));
        
        // Test 3: Word Search II
        char[][] board = {{'o','a','a','n'},{'e','t','a','e'},{'i','h','k','r'},{'i','f','l','v'}};
        String[] words = {"oath","pea","eat","rain"};
        System.out.println("3. Word Search II: " + examples.findWords(board, words));
        
        // Test 4: Replace Words
        List<String> dict = Arrays.asList("cat", "bat", "rat");
        System.out.println("4. Replace Words: " + examples.replaceWords(dict, "the cattle was rattled by the battery"));
        
        // Test 5: Map Sum Pairs
        MapSum mapSum = new MapSum();
        mapSum.insert("apple", 3);
        System.out.println("5. MapSum sum 'ap': " + mapSum.sum("ap"));
        
        // Test 6: Longest Word in Dictionary
        System.out.println("6. Longest Word: " + examples.longestWord(new String[]{"w","wo","wor","worl","world"}));
        
        // Test 7: Search Suggestions System
        String[] products = {"mobile","mouse","moneypot","monitor","mousepad"};
        System.out.println("7. Suggestions 'mouse': " + examples.suggestedProducts(products, "mouse"));
        
        // Test 8: Palindrome Pairs
        System.out.println("8. Palindrome Pairs: " + examples.palindromePairs(new String[]{"abcd","dcba","lls","s","sssll"}));
        
        // Test 9: Word Break II
        System.out.println("9. Word Break II: " + examples.wordBreak("catsanddog", Arrays.asList("cat","cats","and","sand","dog")));
        
        // Test 10: Stream of Characters
        StreamChecker sc = new StreamChecker(new String[]{"cd","f","kl"});
        System.out.println("10. Stream Checker 'a': " + sc.query('a'));
    }
}